File: src/main/java/com/github/zly2006/enclosure/network/EnclosureInstalledC2SPacket.java
package com.github.zly2006.enclosure.network;

import com.github.zly2006.enclosure.ServerMainKt;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
import net.fabricmc.fabric.api.networking.v1.PacketSender;
import net.fabricmc.fabric.api.networking.v1.ServerPlayConnectionEvents;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.fabricmc.loader.api.SemanticVersion;
import net.fabricmc.loader.api.Version;
import net.fabricmc.loader.api.VersionParsingException;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.network.PacketByteBuf;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.network.ServerPlayNetworkHandler;
import net.minecraft.server.network.ServerPlayerEntity;
import net.minecraft.text.Text;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.Map;

import static com.github.zly2006.enclosure.ServerMainKt.MOD_VERSION;

public class EnclosureInstalledC2SPacket implements ServerPlayNetworking.PlayChannelHandler {
    public static final Map<ServerPlayerEntity, Version> installedClientMod = new HashMap<>();

    private EnclosureInstalledC2SPacket() {
        ServerPlayConnectionEvents.DISCONNECT.register((handler, server) ->
                installedClientMod.remove(handler.player));
    }

    public static boolean isInstalled(@Nullable ServerPlayerEntity player) {
        if (player == null) return false;
        return installedClientMod.containsKey(player);
    }

    public static Version clientVersion(ServerPlayerEntity connection) {
        return installedClientMod.get(connection);
    }

    public static void register() {
        EnclosureInstalledC2SPacket listener = new EnclosureInstalledC2SPacket();
        ServerPlayNetworking.registerGlobalReceiver(NetworkChannels.ENCLOSURE_INSTALLED, listener);
    }

    public static void send() {
        PacketByteBuf buf = PacketByteBufs.create();
        buf.writeString(MOD_VERSION.getFriendlyString());
        ClientPlayNetworking.send(NetworkChannels.ENCLOSURE_INSTALLED, buf);
    }

    @Override
    public void receive(MinecraftServer server, ServerPlayerEntity player, ServerPlayNetworkHandler handler, @NotNull PacketByteBuf buf, PacketSender responseSender) {
        Version version;
        try {
            version = Version.parse(buf.readString());
            if (version instanceof SemanticVersion clientVersion && MOD_VERSION instanceof SemanticVersion serverVersion &&
                    clientVersion.getVersionComponent(0) == serverVersion.getVersionComponent(0) &&
                    clientVersion.getVersionComponent(1) == serverVersion.getVersionComponent(1)) {
                ServerMainKt.LOGGER.info(player.getName().getString() + " joined with a matching enclosure client.");
                installedClientMod.put(player, version);

                // send uuid data
                PacketByteBuf buf2 = PacketByteBufs.create();
                NbtCompound compound = new NbtCompound();
                ServerMainKt.byUuid.forEach((uuid, s) -> compound.putUuid(s, uuid));
                buf2.writeNbt(compound);
                compound.put("", new NbtCompound());
                ServerPlayNetworking.send(player, NetworkChannels.SYNC_UUID, buf2);
            } else {
                player.sendMessage(Text.translatable("enclosure.message.outdated", MOD_VERSION.getFriendlyString(), version.getFriendlyString()), false);
            }
        } catch (VersionParsingException ignored) { }
    }
}

File: src/main/java/com/github/zly2006/enclosure/network/EnclosureInstalledC2SPacket.kt
package com.github.zly2006.enclosure.network

import com.github.zly2006.enclosure.MOD_VERSION
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry
import net.fabricmc.fabric.api.networking.v1.ServerConfigurationNetworking
import net.fabricmc.fabric.api.networking.v1.ServerPlayConnectionEvents
import net.fabricmc.loader.api.Version
import net.minecraft.network.PacketByteBuf
import net.minecraft.network.codec.PacketCodec
import net.minecraft.network.packet.CustomPayload
import net.minecraft.server.network.ServerPlayerEntity
import net.minecraft.util.Identifier
import java.util.*

class EnclosureInstalledC2SPacket(
    val version: Version
) : CustomPayload {
    constructor(buf: PacketByteBuf) : this(Version.parse(buf.readString()))

    private fun write(buf: PacketByteBuf) {
        buf.writeString(version.friendlyString)
    }


    override fun getId() = ID

    companion object {
        val installedClientMod: MutableMap<UUID, Version> = HashMap()
        val ID: CustomPayload.Id<EnclosureInstalledC2SPacket?> = CustomPayload.Id(Identifier("enclosure:installed"))
        val CODEC: PacketCodec<PacketByteBuf, EnclosureInstalledC2SPacket?> = CustomPayload.codecOf(
            { obj, buf -> obj!!.write(buf) },
            { buf -> EnclosureInstalledC2SPacket(buf) })

        @JvmStatic
        fun isInstalled(player: ServerPlayerEntity?): Boolean {
            if (player == null) return false
            return installedClientMod.containsKey(player.uuid)
        }

        fun clientVersion(connection: ServerPlayerEntity): Version? {
            return installedClientMod[connection.uuid]
        }

        fun register() {
            PayloadTypeRegistry.configurationC2S().register(ID, CODEC)
            ServerConfigurationNetworking.registerGlobalReceiver<EnclosureInstalledC2SPacket>(ID) { payload, context ->
                if (payload == null) return@registerGlobalReceiver
                installedClientMod[context.networkHandler().debugProfile.id] = payload.version
            }
            ServerPlayConnectionEvents.DISCONNECT.register(ServerPlayConnectionEvents.Disconnect { handler, _ ->
                installedClientMod.remove(handler.player.uuid)
            })
        }

        @JvmStatic
        fun send() {
            ClientPlayNetworking.send(EnclosureInstalledC2SPacket(MOD_VERSION))
        }
    }
}
